### ISSUE TITLE
[Question - can be closed at any moment in time] Java bindings?

### ISSUE THREAD

**Issue by rubyFeedback** [libui-ng/libui-ng#122]
I have been using java lately, in particular swing. Swing is super-ugly though ... libui is also more fun.

Anyway.

https://github.com/andlabs/libui#language-bindings does not show java binding.

Has nobody added them? Is it not possible? Many other languages have bindings to libui ...

How to create such bindings for Java? Is that even possible? I think libui or libui-ng is
in C or C++. But ruby can have bindings, python can have bindings... is it possible with
Java?

**Comment by AndyObtiva** [comment]
Java already has a complete native GUI library called SWT. LibUI currently only supports a tiny fraction of what SWT supports. It’s possible that LibUI was inspired by SWT in fact. 

Anyways, I’m surprised you’re asking this question because a big point of using LibUI is to avoid the slow startup time of the Java Virtual Machine. 

Also, I thought you already heard of SWT and how it’s native (just like LibUI) from Glimmer DSL for SWT (a totally awesome project that can do anything anyone wants from GUI in JRuby running on top of Java): https://github.com/AndyObtiva/glimmer-dsl-swt

**Comment by cody271** [comment]
> Has nobody added them? Is it not possible? Many other languages have bindings to libui ...
>
Good question. Correct, surprisingly no Java bindings exist currently. It is of course very much possible, just hasn't been done yet!

> How to create such bindings for Java? Is that even possible? I think libui or libui-ng is in C or C++. But ruby can have bindings, python can have bindings... is it possible with Java?
>
The best way to create Java bindings is through [JNA](https://en.wikipedia.org/wiki/Java_Native_Access), which is the modern JVM FFI library for creating pure Java bindings to native C code. (And unlike original JNI, no C glue code necessary!)

I have actually thought about JVM bindings, for use from Clojure..

**Comment by cody271** [comment]
> Also, I thought you already heard of SWT and how it’s native (just like LibUI) from Glimmer DSL for SWT (a totally awesome project that can do anything anyone wants from GUI in JRuby running on top of Java): https://github.com/AndyObtiva/glimmer-dsl-swt

Yes for JRuby apps, Glimmer SWT is also a great option. :)

**Comment by AndyObtiva** [comment]
Since Common Lisp (another Lisp other than Clojure) already has libui bindings (https://github.com/jinwoo/cl-ui), I wonder if [ABCL](https://abcl.org/) (Armed Bear Common Lisp) could run libui in the JVM already. It is a Common Lisp implementation that runs in the JVM and can reuse Java libraries from within Lisp. 

In any case, there are other JVM libui bindings like:
- Kotlin: https://github.com/msink/kotlin-libui
- Scala: https://github.com/lolgab/scalaui

I wonder if they're reusable from Java or other JVM languages.

**Comment by msink** [comment]
Existing Kotlin (and probably Scala) binding is not JVM but Native.

**Comment by cody271** [comment]
> I wonder if they're reusable from Java or other JVM languages.

> Existing Kotlin (and probably Scala) binding is not JVM but Native.

Yeah, JVM bindings entirely missing as it is now. Needs to be done with pure Java JNA code at some point.

**Comment by AndyObtiva** [comment]
I'm sorry I was wrong about Kotlin and Scala since they could also be used as JVM languages.
